### DISCUSSION TITLE
변수를 선언한다는 것은 어떤 것을 의미하나요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#18] [변수]
- 변수를 선언한다는 것은 어떤 것을 의미하나요?
  - undefined로 초기화 단계를 거치는 이유는 무엇인가요?

**Comment by pipisebastian** [comment]
> 변수를 선언한다는 것은 어떤 것을 의미하나요?

- **`선언 단계`, `초기화 단계`,** 2단계로 이루어지는데요.
- **`선언 단계`** 는 런타임 이전에 실행되며, 실행 컨텍스트 안에 식별자 정보(`key` - 변수 이름, `value` - 변수 값)가 저장됩니다. (이는 변수 선언이 맨 처음 실행되어, 마치 코드 최상단에 위치한 것처럼 보이기 때문에, 호이스팅이라고도 합니다.)

- **`초기화 단계`** 메모리를 할당하고, `undefined` 로 초기화하는 과정이 일어납니다. 
  - `var` 의 경우, 선언과 동시에 초기화가 일어나, 변수선언문 이전에 변수를 참조하더라도 `undefined`를 가져와 `Reference Error`가 발생하지 않습니다.
  - `let`, `const` 의 경우, 런타임 단계에서 변수 선언문에 도달했을때 초기화가 일어납니다. 
그렇기에 실행컨텍스트에 식별자 정보가 저장된 순간부터, 런타임이후 하나씩 실행하다가 변수 선언문에 도달할때까지, 변수는 초기화가 안된 상태인데요. 그렇기 때문에 이 구간에서는 `Reference Error`가 발생하며, 이 구간을 일시적 사각지대, **TDZ** 이라고 합니다.

> undefined로 초기화 단계를 거치는 이유는 무엇인가요?

확보된 메모리에 쓰레기값이 있을 수도 있기 때문입니다. 이러한 쓰레기값은 예상치못한 결과를 일으킬 수 있기에, 자바스크립트 엔진에서 초기화를 해줍니다.

**Comment by publdaze** [comment]
> - 변수를 선언한다는 것은 어떤 것을 의미하나요?
- 변수 값을 저장하기 위한 메모리 공간을 확보하는 것
> - undefined로 초기화 단계를 거치는 이유는 무엇인가요?
- 해당 메모리 공간에 이전에 저장된 값이 있으면 쓰레기 값이 나올 수 있기 때문에 자바스크립트에서 undefined로 미리 초기화 해줌
